### DISCUSSION TITLE
question about hrv_RSA

### DISCUSSION THREAD

**Discussion by vivile42** [neuropsychology/NeuroKit#763] [Q&A]
Hello, i was checking the P2T RSA function, and from what I see it subtracts the max RR and min RR within a breathing cycle (the cycle is defined from rsp onset to the next so it doesn't differentiate inh exh from what I understand). Shouldn't it be the difference between the min RR during inhalation and max RR during exhalation? I just wanted to make sure I'm understanding it correctly. Also the documentation states that the results are returned in ms but they seem to be in seconds (diff/ sampling rate should give seconds right? then you need to multiply by *1000 to get to ms, otherwise it's really weird because I get RSAs of like 0.04 ms, while 40 makes more sense. )

Thank you a lot in advance for your answer 

(I've copy pasted the section of code I'm referring to)

`` peak algorithm (P2T)"""

    # Find all RSP cycles and the Rpeaks within
    cycles_rri = []
    for idx in range(len(rsp_onsets) - 1):
        cycle_init = rsp_onsets[idx]
        cycle_end = rsp_onsets[idx + 1]
        cycles_rri.append(rpeaks[np.logical_and(rpeaks >= cycle_init, rpeaks < cycle_end)])

    # Iterate over all cycles
    rsa_values = np.full(len(cycles_rri), np.nan)
    for i, cycle in enumerate(cycles_rri):
        # Estimate of RSA during each breath
        RRis = np.diff(cycle) / sampling_rate
        if len(RRis) > 1:
            rsa_values[i] = np.max(RRis) - np.min(RRis) 

    if continuous is False:
        rsa = {"RSA_P2T_Mean": np.nanmean(rsa_values)}
        rsa["RSA_P2T_Mean_log"] = np.log(rsa["RSA_P2T_Mean"])  # pylint: disable=E1111
        rsa["RSA_P2T_SD"] = np.nanstd(rsa_values, ddof=1)
        rsa["RSA_P2T_NoRSA"] = len(pd.Series(rsa_values).index[pd.Series(rsa_values).isnull()])
    else:
        rsa = signal_interpolate(
            x_values=rsp_peaks[~np.isnan(rsa_values)],
            y_values=rsa_values[~np.isnan(rsa_values)],
            x_new=np.arange(len(ecg_period)),
        )

    return rsa``

**Comment by DominiqueMakowski** [comment]
Hi @vivile42 
Yes I think you understood the code correctly. 

> then you need to multiply by *1000 to get to ms, otherwise it's really weird because I get RSAs of like 0.04 ms, while 40 makes more sense. )

Would you be able to make a PR to add that?

> Shouldn't it be the difference between the min RR during inhalation and max RR during exhalation?

I'm not sure, I'm not an expert on RSA (@Tam-Pham ?). It's possible though, if you know of any papers or info that refer to that, please share it with us, it's possible that it has to be fixed indeed :)

**Reply by vivile42** [reply]
Thank you for the answer 
1. yes I can do a pull request with the *1000
2. the references and explanation is already in the documentation of the function on neurokit (I've put a screenshot here below). It does something different from what it's stated.  
<img width="717" alt="image" src="https://user-images.githubusercontent.com/86971275/214045346-b9b5ea7d-b115-4cc6-bcc9-c8eef2ea9d20.png">

the description in the neurokit documentation seems to be taken from the Lewis et al 2011 [article](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3269511/) 
except the word expiration was substituted with breathing cycle 

<img width="901" alt="image" src="https://user-images.githubusercontent.com/86971275/214046890-4bed8af2-8647-464c-ac8f-f82ffa0977f5.png">

Using the max RRI exh - min RRI inn, introduces the issue of how to consider breathing cycles in which the RRI are longer during inhaling than exh (they can be for example set to 0), in the lewis paper they discuss 2 different methods.

**Reply by DominiqueMakowski** [reply]
Indeed, that seems correct. In that case let's do all the fixes, + fix the documentation. Maybe we could streamline the documentation to be clearer and less copy-paste from that paragraph :) Let me know if you need any help in the PR

**Reply by vivile42** [reply]
I've run the check code commands the way it's said in the contribution guide and it slightly changed a lot of code that was already there, should I send the pull request as it is or since it's just a * 1000  should I skip the reformatting to avoid messing with other parts? 

Also I think I will start fixing the code for the P2T mean anyway since I need it soonish for my project,  I can try do a PR with my changes once I've tested it and then you can decide whether to keep it or change it differently. Let me know otherwise I'll just do it locally.

**Reply by DominiqueMakowski** [reply]
Yes don't bother too much with all the checks, you can just push the changes that you have (*1000 + P2T) and we'll see from there
